test/common/upstream/health_check_fuzz.cc

```diff
@@ -245,17 +249,24 @@ void TcpHealthCheckFuzz::respond(std::string data, bool last_action) {
   // turn on interval and turn off timeout, but for tcp it doesn't if the data doesn't match. If the
   // response doesn't match, it only sets the host to unhealthy. If it does match, it will turn
   // timeout off and interval on.
-  if (!reuse_connection_ && !last_action && interval_timer_->enabled_) {
-    expectClientCreate();
-    interval_timer_->invokeCallback();
+  if (!reuse_connection_ && interval_timer_->enabled_) {
```

Comment:
Why is it ok to trigger the interval even if it is the last action (while previously it wasn't)?

Reply:
Two things: 1. In response to your comment, I personally found it cleaner to have the data needed to respond with clearly declared as an argument in the function call. It gives a clear distinction in the header file as well for a reader what the difference is, since that is what the main difference is across all three types. However, I can totally change it.

2. Yeah, this is a completely valid point. Honestly, the last action stuff isn't needed, and it's just put for efficiency. The fuzzer right now isn't consistent, I believe that respond doesn't account for last action like the other actions. I can make the respond methods also take last_action into account.

Reply:
For the sake of consistency, I'll add last action to the responds.

Reply:
If you are certain that the last_action isn't needed, should it be discard?
I'm assuming that it will make the code easier to follow, but don't have a strong opinion whether it should be kept or not.

Reply:
@asraa Yup, valid opinion on both sides, me and Asra have already talked about this, let's get her input as well :)!

Reply:
Hey Adi, me and Asra talked about this in standup, and decided to keep the last_action. First, the removal of last_action would change zero functionality in the code. Essentially, the high level design of my fuzzer is to explicitly invoke interval timer in certain cases after events, so that you don't have to wait for the fuzzer to generate an invoke interval timer to proceed with events. https://github.com/envoyproxy/envoy/blob/master/source/common/upstream/health_checker_impl.cc#L252 for example, this is what happens with HTTP health checking if you invoke interval timer. It creates (client?), and stream, turns on timeout timer and turns off interval timer, and also sends out a request. Thus, removing last_action is logically and functionally no different to keeping it, as if it does trigger a crash, it would trigger in a generated event with that action not as the last action. Thus, it doesn't add or retract any crashes. Removing it would leave dangling requests at the end of each iteration.

I'll add a comment explaining this, and add last_action to all my responds. Thanks for starting discussion about it :).

Reply:
Sounds good to me, thanks!